### PR TITLE
i318: merge rather than replace group list on load of teams.tsv

### LINKED ISSUE #318: teams.tsv import unconditionally clears group - it shouldn't.
**Describe the issue**: 

When teams tsv is imported (ICPC - Import CMS tsv files) the group field is set to empty/null.

It would be helpful if any future NPE or other Exceptions show a dialog to the user,
if variable useConfirmGUI is true (meaning the load is done via GUI rather than command line/no GUI)

**To Reproduce**: please describe, in order, the steps to reproduce the behavior:

Start a server with a sample with a teams.tsv, --load samps\contests\valtest
Start Admin
Select Accounts tab, note all teams are assigned groups
ICPC -> Import CMS tsv files
Select \contests\valtest\config\teams.tsv, answer Yes
Select Accounts tab

**Expected behavior**: 
Team account information should not change

**Actual behavior**: 
No teams are assigned a group

NPE logged into log

**Environment**: 
All, tested using
Version 9.8build 20220223 (Wednesday, February 23rd 2022 23:46 UTC) build 6215~develop
Java version 1.8.0_121
OS: Windows 10 10.0 (amd64)

**Log Info**:  

220228 082208.548|INFO|AWT-EventQueue-0|loadTSVFiles|Exception in loadTSVFiles()
java.lang.NullPointerException: null
    at java.io.File.<init> (File.java:277)
    at edu.csus.ecs.pc2.core.imports.LoadICPCTSVData.checkFiles (LoadICPCTSVData.java:253)
    at edu.csus.ecs.pc2.core.imports.LoadICPCTSVData.loadFiles (LoadICPCTSVData.java:85)
    at edu.csus.ecs.pc2.core.imports.LoadICPCTSVData.loadFiles (LoadICPCTSVData.java:72)
    at edu.csus.ecs.pc2.ui.ICPCLoadPane.loadTSVFiles (ICPCLoadPane.java:434)
    at edu.csus.ecs.pc2.ui.ICPCLoadPane$5.actionPerformed (ICPCLoadPane.java:372)

**Screenshots**:  

**Additional context**:

### PR DESCRIPTION
**Pull request by lane55**
### Description of what the PR does

Groups were being blanked out on Accounts list on load  of teams.tsv and groups.tsv.
Now merges model's groups and loaded groups from groups.tsv.

### Issue which the PR fixes

#318 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)

Microsoft Windows [Version 10.0.19044.1586]

java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

tldr;   Start server with sample contest that has teams.tsv, groups.tsv, 
   load the same teams.tsv using ICPC tab -> Import CMS tsv files
   view Accounts, and confirm that groups are blanked out.

Load sample contest mini
 pc2server --login s --contestpassword contest  --load mini
Start Admin 
note that groups are populated for accounts
ICPC tab -> Import CMS tsv files 
load either groups.tsv or teams.tsv from samps\contests\mini\config
 
Expected

In Accounts tab, groups are present.

Actual:

All team groups are blank.

### CODE REVIEW

File: src/edu/csus/ecs/pc2/core/imports/LoadICPCTSVData.java [hunk 1]
```diff
@@ -227,6 +225,8 @@ protected void updateGroupsAndAccounts(IInternalContest inContest, List<Group> g
             
             accountList.set(i, existingAccount);
             i ++;
+            
+
```
Comment (clevengr) [inline src/edu/csus/ecs/pc2/core/imports/LoadICPCTSVData.java:229]: I can understand the addition of a _single_ blank line for improved readability, but personally I think adding TWO blank lines just makes it _harder_ to read the code.   [This is not a request for a change; just an observation...]

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java [hunk 2]
```diff
@@ -319,17 +329,57 @@ public static Group[] loadGroups(String filename) throws Exception {
...
-            groups[groupAccount] = group;
-            groupAccount++;
+            
+            groupList.add(group);
         }
-
+        
+        /**
+         * Add groups from authoritativeGroups if missing from groupList.
+         */
```
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:346]: I do not understand the purpose of, or the need for, the following for-loop.  The comment says that it's to add to `groupList` any groups that were in `authoritativeGroups` but are missing from `groupList`.  However, the new code added to the previous for-loop (just above here) obtains a group from the input file, then looks to see if that group is in `authoritativeGroups` (using `Group group = findGroupById(authoritativeGroups, number);`).  It then adds the group (either the one found in `authoritativeGroups` or else a newly-created group) to `groupList`.   Therefore, at THIS point in the code _every_ group found in `authoritativeGroups` MUST also now be in `groupList` -- so what's the point in the entire for-loop following this comment?
Or am I missing something?

Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:346]: BTW, I would not block approval for this -- but I don't like having chunks of code that are useless.   (Unless I'm wrong about what's going on, in which case please explain how I'm wrong...)

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java [hunk 3]
```diff
@@ -319,17 +329,57 @@ public static Group[] loadGroups(String filename) throws Exception {
...
         return groups;
     }
-    
+
+    /**
+     * Search and return group if found.
+     * 
+     * @param authoritativeGroups list of groups
+     * @param groupCMSId group (CMS
+     * @return null if not found, else the group
```
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:372]: What happened to the rest of the comment here?

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java [hunk 4]
```diff
@@ -376,4 +427,6 @@ public static HashMap<Integer, String> loadPasswordsFromAccountsTSV(String filen
         }
         return passwordMap;
     }
+
+
```
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:431]: Again... why two blank lines?

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java [hunk 5]
```diff
@@ -111,5 +115,88 @@ private void checkIfGroupsAssigned(List<Account> accountList) {
...
+    private String getTestSampleContestConfigFile(String contestName, String filename) {
+        return getTestSampleContestDirectory(contestName) + File.separator + IContestLoader.CONFIG_DIRNAME + File.separator + filename;
+    }
+
+    /**
+     * Test loading a teams.tsv twice.
+     * 
+     * The 2nd load of teams and groups, in particular groups, caused all accounts' groups to be "empty"
+     * on the Accounts tab.  See 
+     *
```
Comment (clevengr) [inline test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:144]: See.... what?

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java [hunk 6]
```diff
@@ -111,5 +115,88 @@ private void checkIfGroupsAssigned(List<Account> accountList) {
...
+
+        IInternalContest contest = loadSampleContest(null, "mini");
+        assertNotNull(contest);
+        IInternalController controller = new SampleContest().createController(contest, true, false);
+        
+        LoadICPCTSVData loader = new LoadICPCTSVData();
+        loader.setContestAndController(contest, controller);
+        
+        boolean loaded = loader.loadFiles(groupsFilename, false, false);
+        assertTrue("Expecting min contest loaded", loaded);
```
Comment (clevengr) [inline test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:160]: I think maybe the word in the message was supposed to be "mini", not "min"?

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java [hunk 7]
```diff
@@ -111,5 +115,88 @@ private void checkIfGroupsAssigned(List<Account> accountList) {
...
+         * Groups from contest/model, the authoritative groups from model.
+         */
+        Group[] modelGroups = contest.getGroups();
+
+        assertEquals("Number of groups", 12, modelGroups.length);
+        assertEquals("Number of accounts", 151, accounts.length);
+
+        /**
+         * Second load of groups from groups.tsv
+         */
```
Comment (clevengr) [inline test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:183]: This is a repeat of a comment above; that doesn't seem accurate.

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java [hunk 8]
```diff
@@ -111,5 +115,88 @@ private void checkIfGroupsAssigned(List<Account> accountList) {
...
+        // Load groups from file, merge with modelGroups list.
+        Group[] mergedGroups = ICPCTSVLoader.loadGroups(loader.getGroupsFilename(), modelGroups);
+
+        Account[] accounts2 = ICPCTSVLoader.loadAccounts(loader.getTeamsFilename());
+
+        assertEquals("Number of groups", 12, mergedGroups.length);
```
Comment (clevengr) [inline test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:189]: This doesn't seem like a valid (or at least, not a very good) test of merging -- the group sizes (12) are the same before and after the merge. Unless I've missed something which makes them not the same?  If so, please illuminate me...

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java [hunk 9]
```diff
@@ -111,5 +115,88 @@ private void checkIfGroupsAssigned(List<Account> accountList) {
...
+        assertEquals("Number of accounts", 151, accounts2.length);
+
+        for (int i = 0; i < modelGroups.length; i++) {
+            assertEquals("(" + i + ") Expecting same group element id for group " + modelGroups[i], modelGroups[i].getElementId(), mergedGroups[i].getElementId());
+        }
+
+        // check that all accounts assigned groups in the model/contest
+        for (Account account : accounts2) {
+            Group group = contest.getGroup(account.getGroupId());
```
Comment (clevengr) [inline test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:198]: `account.getGroupId()` can return null (if no Group is assigned).  I don't know if that's a problem here or not...

**Review by clevengr — COMMENTED** [review]
I've made a number of comments.  None are overwhelming, but I think it's worth at least having you review them, and "Resolve" them if you think appropriate.

File: src/edu/csus/ecs/pc2/core/imports/LoadICPCTSVData.java (hunk 1 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/core/imports/LoadICPCTSVData.java:229]: noted.

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 8 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:189]: This test ensures that after the merge there are the proper, 12, number of groups.
If there are other than 12 then the merge has failed.

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 9 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:198]: It can, and if it does then this unit test will fail... and the cause can be fixed/researched.

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 3 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:368]: java doc updated.

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 2 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:342]: The comment is correct.     Merge may imply that the number of groups in the input file are identical to
the number/data for the existing authoritativeList.   That would be a bug.   There may be more or different
groups in the authoritative list than there are in the input file.    The merge is a union of groups in
the input file and the authoritativeList (model).   For example there could be 12 authoritativeList
groups and only 4 groups in the input file.  There should be a minimum of 12 groups output.  Without
the second loop there would be 4 groups output.

Logic is:
The top loop merges the authoritative/existing group data and the groups from the input file.  The comment is:
   "Add groups from authoritativeGroups if missing from groupList."
The bottom loop adds any missing groups from the authoritative list into the method's output groupList

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 4 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:427]: Stuff happens.   I would like to focus on non-whitespace and non-format related actual code work/review.

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 5 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:144]: fixed comment, added issue url

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 6 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java]: updated to use variable for contestName, pushed.

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 2 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:342]: Ok, it took me a while but now I get why the second loop is needed.  However, it took me a LONG time to understand what the second loop is actually doing, due to very bad naming (see below).  I believe the code works -- but it's really confusing to the reader (again, see below).

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java [hunk 10]
```diff
@@ -319,17 +329,57 @@ public static Group[] loadGroups(String filename) throws Exception {
...
+        /**
+         * Add groups from authoritativeGroups if missing from groupList.
+         */
+        for (Group aGroup : authoritativeGroups) {
+            
+            Group missingGroup = null; 
+            
+            for (Group glGroup : groupList) {
+                if (aGroup.getGroupId() == glGroup.getGroupId()) {
+                    missingGroup = glGroup;
```
Comment (clevengr) [inline src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:349]: `missingGroup` is a VERY bad/misleading name for this variable.  In fact, when a value is assigned to `missingGroup`, that  value is a Group which is NOT missing (from `groupList`).  The reversed logic of this is quite confusing (although I'll concede that the binary works...)

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 4 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:427]: `I would like to focus on non-whitespace and non-format related actual code work/review.`  Sure, I get that.  But if I'm reviewing code I'm going to comment where I think there's an issue even if it's trivial; failure to do so means things creep into code that make it harder in the future (in some, very weak in this case, sense).  I did note that I certainly wouldn't refuse to approve such code...

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 8 above)
Comment (clevengr) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:183]: Right, but my point was:  what about the case where 12 is NOT the proper number?  E.G. when a new Group is being added?  It seems like THAT case is a much more substantial test of whether a "merge" worked correctly.

**Review by clevengr — APPROVED** [review]

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 7 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java]: removed.

File: test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java (hunk 8 above)
Comment (lane55) [reply test/edu/csus/ecs/pc2/core/imports/LoadICPCTSVDataTest.java:183]: You asked about why it was valid, not what a better test would be.   Sure there are bunches of tests
that could be added, I added that test for the reasons above.

File: src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java (hunk 10 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:349]: Suggest a less misleading name, rather than just comment.   I used that name thinking at the time that
it made sense (or more likely the code section was rewritten keeping that name).

Comment (clevengr) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:349]: Well, since what this loop is doing is checking, for each Group in `authoritativeGroups`, whether that group is found in `groupList` (and if NOT, it is adding the Group from `authoritativeGroups` into `groupList`, a "better" name might be "found".

However, the real problem is the confusion in what this loop is trying to accomplish vs. the code it uses to accomplish that.  What it is trying to accomplish (according to the comment at the top, and, unless I'm mistaken, in the code itself) is to add to `groupList` any Group which is currently in `authoritativeGroups` but is NOT already in `groupList`.  I think a better solution would be to replace the entire loop with something like:
```
       for (Group aGroup : authoritativeGroups) {
           if (! groupList.contains(aGroup) {
              groupList.add(aGroup);
              break;
          }
       }
```
This is much simpler and clearer, and I believe it accomplishes what is intended (although I would certainly expect you to review that assertion :)  )

Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java:349]: So no change in the name of the variable, ok.

Comment (lane55) [reply src/edu/csus/ecs/pc2/imports/ccs/ICPCTSVLoader.java]: pushed fix.
